lib/provider.js

```diff
@@ -0,0 +1,209 @@
+    this._throw(errors.MethondUnimplemented, 'Provider.workerInfo()');
+  }
+
+  /**
+   * Any code which is required to be run by this Provider must only be
+   * initiated by this method.  If there's any taskcluster-lib-iterate loops to
+   * run, this is where they should be initiated.  Once the returned promise is
+   * resolve, the Provider must be fully working.
+   *
+   * This method must not start anything which does actual provisioning, but
+   * rather is intended to do things like update pricing information or
+   * maintaining state in the underlying api.
+   */
+  async initiate() {
+  }
+
+  /**
+   * Terminate any code which was started by .initiate();
+   */
+  async terminate() {
+  }
+
+  /**
+   * This method must return a list of valid bids based on the results of an
+   * internal evaluation of the workerConfiguration.  It should consider all
+   * satisfiers which might be relevant in the execution.
+   *
+   * This method must do worker configuration evaluation as it sees fit, but
+   * will be passed a workerConfiguration which only allows access to
+   * providerData and workerType in the evaluation results.
+   *
+   * The number of bids returned is at the discretion of the Provider author.
+   * Consideration should be taken to not do bad things, like return 10,000,000
+   * bids for a demand of 1.
+   *
+   * It is the provider's responsibility to ensure that the bid is valid for the
+   * provider.  Things like checking structural validity must be handled
+   * within the provider, since the rest of the worker manager will treat
+   * these values as opaque.  The provider must also ensure that the
+   *
+   * If this provider must track maximum instance counts, it must track this
+   * internally as this is considered to be a provider concern
+   */
+  async proposeBids({workerType, workerConfiguration, demand}) {
+    this._throw(errors.MethondUnimplemented, 'Provider.proposeBids()');
+  }
+
+  /**
+   * Submit the provided set of bids.  If this provider internally tracks the
+   * state of pending bids, they should be marked as resolved.
+   *
+   * This method should not throw because the bids are rejected internally.
+   * Each provider should try to retry failed bids, and mark them internally
+   * against any reliability scores used to determine whether to offer similar
+   * bids in the future.
+   *
+   * No return value is expected.  This function should only throw an error if
+   * the bids parameter is malformed.
+   */
+  async submitBids({bids}) {
+    this._throw(errors.MethondUnimplemented, 'Provider.submitBids()');
+  }
+
+  /**
+   * Mark unused bids as rejected.  This is to account for a system like
+   * provisioning a fixed size pool in case a provider wanted to do resource
+   * counting so that it never suggests the same resource for more than one
+   * bid.  Since the majority of providers will not implement this behaviour, a
+   * default non-operation implementation is provided.
```

Comment:
Is this maintained over service restarts?  Or is this just within a single process?

Reply:
that is up to the provider to decide.  Providers will be allocated storage in the worker-manager's data storage, so if they wish to persist/share this information, it is their prerogative